keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -0,0 +1,631 @@
+
+If a service is configured for an endpointFamily of "ipv4" ("ipv6"), then endpoints for that port for the service will not include any IPv6 (IPv4) addresses that have been allocated to backend pods for the service.
+
+If a service is configured with an endpointFamily of "dual-stack", then both IPv4 and IPv6 endpoints will be created, but the service will only have a single service IP allocated (with a family that matches the cluster's configured service CIDR).
+
+If a service is exposed via nodePort, and the service is configured with an endpointFamily of "dual-stack", then iptables will be configured for both families, allowing both IPv4 and IPv6 forwarding.
+
+#### Modifying Consumers of EndpointAddress API to Use Dual-Stack Addresses
+Any clients that consume the [Kubernetes EndpointAddress API](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#endpointaddress-v1-core) will need to be modified to read the new plural IPs field, if they are to take advantage of the dual-stack feature. This proposal will include dual-stack updates to the following consumers of Kubernetes endpoints:
+- kube-proxy, iptables mode
+- kube-proxy, IPVS mode
+- Nginx Ingress controller
+Other users/consumers of this API may be updated to dual-stack over time, but that work is considered outside of the scope of this proposal.
+
+#### 'kubectl get endpoints' Command Display for Dual-Stack Backend Pods
+The 'kubectl get endpoints ...' command display should be changed so that:
+- IPv4/IPv6 endpoint address pairs appear within curly braces.
+- If dual-stack endpoint address pairs are present, then each pair should be printed on its own line as show below.
+Example:
+```
+       kube-master# kubectl get endpoints
+       NAME            ENDPOINTS                             AGE
+       kubernetes      {[fd00::100]:6443,10.0.0.2:6643}      15m
+       nginx-service   {[fd00:db8:1::2]:80,192.168.1.3:80}   20m
+                       {[fd00:db8:2::2]:80,192.168.2.3:80}   20m
+       kube-master#
+```
+
+#### 'kubectl describe service' Command Display for Dual-Stack Backend Pods
+The 'kubectl describe service ...' command display should be changed so that:
+- IPv4/IPv6 endpoint address pairs appear within curly braces.
+- If dual-stack endpoint address pairs are present, then each pair should be printed on its own line as show below.
+Example:
+```
+       kube-master# kubectl describe service nginx-service
+       .
+       .
+       .
+       Endpoints:   {[fd00:db8:1::2]:80,192.168.1.3:80}
+                    {[fd00:db8:2::2]:80,192.168.2.3:80}
+       .
+       .
+       .
+       kube-master#
+```
+
+### kube-proxy Operation
+
+Kube-proxy will be modified to drive iptables and ip6tables in parallel. This will require the implementation of a second "proxier" interface in the Kube-Proxy server in order to modify and track changes to both tables. This is required in order to allow exposing services via both IPv4 and IPv6, e.g. using Kubernetes:
```

Comment:
Could you clarify the "second proxier interface"? Would it be cleaner to just build both v4 and v6 iptables rules while processing each ServicePort and then do two restores at the same time (one for v4 and one for v6) when the service map is done being processed?

Reply:
I think "Two proxier instances" would be a better description. The proxier is already instantiated either for ipv4 or ipv6. In dual-stack the cleanest way is to instantiate both. Perhaps with a "meta proxier" in between hiding the instances and calls one or the other or both depending on setup. As a side remark, the ipv4 and ipv6 operations are disjunkt and can be executed in paralell.

If not having two proxier instances you would get something like;
```
  if ipv4Enabled {
    // Setup ipv4 stuff
  }
  if ipv6Enabled {
    // Setup ipv6 stuff
  }
```
all over the place, a construct that should be avoided. @dcbw I don't think this is what you suggest but just to claeify.

Reply:
As above, I think dual-stack kube-proxy can push to phse 2 (or 1.5 -- after all the work and testing for phase 1)